File: ABC179/E_Sequence_Sum/submission3.java
// AC

import java.util.Scanner;

class Main {
  public static void main(String[] args) {
    var sc = new Scanner(System.in);
    long n = sc.nextLong();
    int x = sc.nextInt();
    int m = sc.nextInt();

    long answer = 0;
    int x2 = x;
    long i = 0;

    var s = new int[m];
    var sb = new int[(m + 31) / 32];
    for (; ; i++) {
      if ((sb[x2 / 32] & (1 << (x2 % 32))) != 0) {
        int j = (int)i - 1;
        for (; j >= 0; j--) {
          if (s[j] == x2) {
            break;
          }
        }
        long answer2 = 0;
        for (int k = j; k < i; k++) {
          answer2 += s[k];
        }
        answer += (n - i) / (i - j) * answer2;
        i += (n - i) / (i - j) * (i - j);
        break;
      }
      if (i >= n) break;
      answer += x2;
      s[(int)i] = x2;
      sb[x2 / 32] |= 1 << (x2 % 32);
      x2 = (int)((long)x2 * x2 % m);
    }
    for (; ; i++) {
      if (i >= n) break;
      answer += x2;
      x2 = (int)((long)x2 * x2 % m);
    }

    System.out.println(answer);
  }
}

File: ABC178/C_Ubiquity/submission3.scala
// AC

object Main extends App {
  val m = 1_000_000_007;

  def pow(a: Int, b: Int): Int = {
    @scala.annotation.tailrec
    def sub(a: Long, b: Int, result: Long): Int = {
      if (b == 0) {
        result.toInt;
      } else {
        val result2 = if (b % 2 == 1) result * a % m else result;
        sub(a * a % m, b / 2, result2);
      }
    }
    sub(a, b, 1);
  }

  val sc = new java.util.Scanner(System.in);
  val n = sc.nextInt();

  val answer = (2L * m + pow(10, n) - 2L * pow(9, n) + pow(8, n)) % m;
  println(answer);
}
